### PR TITLE
Add Experimental State Visualizer component 

### PR DESCRIPTION
**Pull request by CDiaz-MS**
## Overview
The State Visualizer component adds animations to an object based on the states defined in a linked Interactive Element component.  This component creates animation assets, places them in the MixedRealityToolkit.Generated folder and enables
simplified animation keyframe setting through adding animatable properties to a target game object.
To enable animation transitions between states, an Animator Controller asset is created and a default state machine
is generated with associated parameters and transitions. 

This is an experimental component and this branch is going into the interactive_element feature branch.

This is a Unity 2019 feature because SerializeReference is utilized.

### How to use the State Visualizer component 
1. Create a sphere and attach an Interactive Element component and a State Visualizer component.   Select "Generate New Animation Clips" to create the state machine, animation controller asset, and animation clip assets for each state

| Inspector view | Generated state machine in the Animator Window|
|---|---|
|AddStateVisualizer| image|

3. In the Focus state, select "Add Target" and add the current sphere as the target
AddAnimationTarget

4. Add the Color animatable property and select a color to transition to and close the fold out to set the key frames

| Inspector view | Animation clip view in the Animation window|
|---|---|
|SetColorKeyframes|SetColorKeyframesAnimationView|

6. Press play and see the transition of color when the object is in Focus. 
PlayModeColorChangeView

## To Do
- [x]  ~~Move to experimental~~ This is a Unity 2019 feature that will be put in a directory with an assembly definition that has a define constraint for Unity 2019.3 or above in another PR 
- [x] Add experimental tags in the inspector

### CODE REVIEW

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimationTarget.cs [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+            set
+            {
+                if (IsTargetObjectValid(value))
+                {
+                    target = value;
+                }
+                else
+                {
+                    target = null;
+                    Debug.LogError("The Target property can only be a child of this game object, the target was set back to null");
```
Comment (keveleigh) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimationTarget.cs:36]: Should this set it to `null`, or just leave it as whatever it's currently set to (which might not be `null`!)?

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimationTarget.cs [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        /// <summary>
+        /// List of animatable properties for the target game object.  Scale and material color are examples of animatable properties.
+        /// </summary>
+        public List<IStateAnimatableProperty> StateAnimatableProperties
+        {
+            get => stateAnimatableProperties;
+            internal set => stateAnimatableProperties = value;
+        }
+
+        public void SetKeyFrames(AnimationClip animationClip)
```
Comment (keveleigh) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimationTarget.cs:54]: nit: missing some summaries on public methods

**Comment by julenka** [comment]
Following[ "Experimental Features should live in a separate folder"](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-live-in-a-separate-folder), make sure to put all your scripts, prefabs, and scenes into an experimental folder, following recommended structure.

File: Assets/MRTK/SDK/Editor/Inspectors/UX/StateVisualizer/StateVisualizerInspector.cs [hunk 3]
```diff
@@ -0,0 +1,402 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License
+
+using Microsoft.MixedReality.Toolkit.UI.Interaction;
+using Microsoft.MixedReality.Toolkit.Utilities.Editor;
+using UnityEditor;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.Editor
```
Comment (julenka) [inline Assets/MRTK/SDK/Editor/Inspectors/UX/StateVisualizer/StateVisualizerInspector.cs:9]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimatableProperty.cs [hunk 4]
```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License
+
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimatableProperty.cs:4]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimationTarget.cs [hunk 5]
```diff
@@ -0,0 +1,195 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License
+
+using Microsoft.MixedReality.Toolkit.Utilities;
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimationTarget.cs:9]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ColorStateAnimatableProperty.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License
+
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ColorStateAnimatableProperty.cs:6]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/PositionOffsetStateAnimatableProperty.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License
+
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/PositionOffsetStateAnimatableProperty.cs:7]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ScaleOffsetStateAnimatableProperty.cs [hunk 8]
```diff
@@ -0,0 +1,57 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License
+
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ScaleOffsetStateAnimatableProperty.cs:6]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ShaderProperties/ShaderColorStateStateAnimatableProperty.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License
+
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ShaderProperties/ShaderColorStateStateAnimatableProperty.cs:6]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ShaderProperties/ShaderFloatStateAnimatableProperty.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License
+
+using System;
+using System.Collections.Generic;
+using UnityEditor;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ShaderProperties/ShaderFloatStateAnimatableProperty.cs:9]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ShaderProperties/ShaderStateAnimatableProperty.cs [hunk 11]
```diff
@@ -0,0 +1,61 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License
+
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ShaderProperties/ShaderStateAnimatableProperty.cs:6]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ShaderProperties/ShaderVectorStateAnimatableProperty.cs [hunk 12]
```diff
@@ -0,0 +1,77 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License
+
+using System;
+using System.Collections.Generic;
+using UnityEditor;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ShaderProperties/ShaderVectorStateAnimatableProperty.cs:9]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/IStateAnimatableProperty.cs [hunk 13]
```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License
+
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/IStateAnimatableProperty.cs:6]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/StateAnimatableProperty.cs [hunk 14]
```diff
@@ -0,0 +1,121 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License
+
+using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateAnimatableProperties/StateAnimatableProperty.cs:8]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateContainer.cs [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+#if UNITY_EDITOR
+using UnityEditor.Animations;
+#endif
+using UnityEngine;
+
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateContainer.cs:14]: Make sure that all experimental classes are in a separate experimental namespace, following the [guide here](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/ExperimentalFeatures.html#experimental-code-should-be-in-a-special-namespace)

File: Assets/MRTK/SDK/Editor/Inspectors/UX/StateVisualizer/StateVisualizerInspector.cs [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+            {
+                RenderStateContainers();
+            }
+
+            RenderEndingButtons();
+
+            serializedObject.ApplyModifiedProperties();
+        }
+
+        private void RenderExperimentalWarning()
```
Comment (julenka) [inline Assets/MRTK/SDK/Editor/Inspectors/UX/StateVisualizer/StateVisualizerInspector.cs:62]: Can you re-use the ExperimentalDrawer class rather that re-implementing the HelpBox rendering here? Would be good to reduce code duplication so that we don't need to change this message below in multiple places if we decide to change the message in the future.

File: Assets/MRTK/SDK/Editor/Inspectors/UX/StateVisualizer/StateVisualizerInspector.cs [hunk 17]
```diff
@@ -0,0 +1,402 @@
...
+
+                serializedObject.ApplyModifiedProperties();
+
+                SerializedProperty newAnimationTarget = animationTargetList.GetArrayElementAtIndex(animationTargetList.arraySize - 1);
+                newAnimationTarget.FindPropertyRelative("target").objectReferenceValue = null;
+
+                SerializedProperty stateAnimatablePropertiesList = newAnimationTarget.FindPropertyRelative("stateAnimatableProperties");
+
+                // Clear the new list
+                for (int k = 0; k < stateAnimatablePropertiesList.arraySize; k++)
```
Comment (julenka) [inline Assets/MRTK/SDK/Editor/Inspectors/UX/StateVisualizer/StateVisualizerInspector.cs:251]: Could this be why you were seeing that null reference error that you ran into when deleting a property? Try iterating through the list backwards, or see if stateAnimatablePropertiesList has a method to clear the array.

If that doesn't fix it, treat this comment as a reminder to fix that null reference error you hit when trying to remove a property from an animation that had just one element in it :)

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimationTarget.cs [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+                animatableProperty.RemoveKeyFrames(animationClip);
+            }
+        }
+
+        private IStateAnimatableProperty GetAnimatableProperty(string animatablePropertyName)
+        {
+            return StateAnimatableProperties.Find((prop) => prop.AnimatablePropertyName == animatablePropertyName);
+        }
+
+        private bool IsTargetObjectValid(GameObject target)
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimationTarget.cs:78]: See my previous comment. I think you can reuse `TransformExtensions.FindAncestorComponent<StateVisualizer>(target, includeSelf=true)` rather than needing to re-implement this method. Also try to factor out IsTargetObjectValid into a common class to avoid code duplication.

File: Assets/MRTK/SDK/Editor/Inspectors/UX/StateVisualizer/StateVisualizerInspector.cs [hunk 19]
```diff
@@ -0,0 +1,402 @@
...
+                if (instance.EditorAnimatorController != null)
+                {
+                    RenderSyncStatesButton();
+                }
+            }
+        }
+
+        // A target game object is one that is itself or a child of the root
+        private bool IsTargetObjectValid(GameObject targetObj)
+        {
```
Comment (julenka) [inline Assets/MRTK/SDK/Editor/Inspectors/UX/StateVisualizer/StateVisualizerInspector.cs:374]: It looks like this method is just checking if a StateVisualizerComponent is present on this object, or on an ancestor. It looks like you can actually reuse `TransformExtensions.FindAncestorComponent<StateVisualizer>(this Transform startTransform, includeSelf=true)` rather than needing to re-implement this method. 

example of how you would use it: `return targetObj.transform.FindAncestorComponent<StateVisualizer>(true);`

Also, I noticed that you have the same method in another .cs file in this PR. Consider trying to factor that method out into one place to avoid code duplication.

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimationTarget.cs [hunk 20]
```diff
@@ -0,0 +1,195 @@
...
+                if (animatableProp.AnimatablePropertyName == animatablePropertyTypeName)
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        private int GenerateIDShaderProperty()
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimationTarget.cs:181]: Super minor, more as an FYI. I think you can use the LINQ [.Count()](https://docs.microsoft.com/en-us/dotnet/api/system.linq.enumerable.count?view=net-5.0) function to do the same thin. Something like `StateAnimatableProperties.Count((x => x is ShaderStateAnimatableProperty))`

**Review by julenka — COMMENTED** [review]
Some initial comments, all the way up to but not including StateContainer.cs. Biggest comment is regarding making sure that this experimental feature is in the proper folder and namespace.

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateVisualizer.cs [hunk 21]
```diff
@@ -0,0 +1,545 @@
...
+using System.Runtime.CompilerServices;
+using UnityEditor;
+
+#if UNITY_EDITOR
+using UnityEditor.Animations;
+#endif
+using UnityEngine;
+
+[assembly: InternalsVisibleTo("Microsoft.MixedReality.Toolkit.SDK.Editor")]
+namespace Microsoft.MixedReality.Toolkit.UI.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/StateVisualizer.cs:17]: Reminder to put in experimental namesace

**Review by julenka — COMMENTED** [review]
Read through the rest of the files, they seem all right to me.

File: Assets/MRTK/SDK/Editor/Inspectors/UX/StateVisualizer/StateVisualizerInspector.cs (hunk 17 above)
Comment (CDiaz-MS) [reply Assets/MRTK/SDK/Editor/Inspectors/UX/StateVisualizer/StateVisualizerInspector.cs]: I fixed this issue, another index value should have been the parameter for removing the keyframes

File: Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimationTarget.cs (hunk 1 above)
Comment (CDiaz-MS) [reply Assets/MRTK/SDK/Features/UX/Scripts/StateVisualizer/AnimationTarget.cs]: This property needs to be set to null because a target object that does not have a state visualizer attached or is a child of an object with state visualizer attached will never be valid.

File: Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/InteractiveElement/BaseInteractiveElement.cs [hunk 22]
```diff
@@ -24,6 +24,7 @@ public abstract class BaseInteractiveElement :
         IMixedRealityPointerHandler,
         IMixedRealitySpeechHandler
     {
+        [Experimental]
```
Comment (julenka) [inline Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/InteractiveElement/BaseInteractiveElement.cs:27]: Was this change intentional in this PR? Reminder for if you move BaseInteractiveElement to experimental: move it to the experimental folder, and experimental namespaces as well

File: Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/InteractiveElement/CompressableButton.cs [hunk 23]
```diff
@@ -9,7 +9,7 @@
 using UnityEngine.Events;
 using UnityEngine.Serialization;
 
-namespace Microsoft.MixedReality.Toolkit.UI.Interaction
+namespace Microsoft.MixedReality.Toolkit.Experimental.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/InteractiveElement/CompressableButton.cs:12]: Might want to separate moving interactiveelement to experimental in a separate PR. Also, reminder to also move all files for interactive element into experimental and to add experimental attribute. Since that will be a lot of code changes, I think it would be better to do that move in a separate PR rather than here.

File: Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/InteractiveElement/Events/EventConfigurations/BaseInteractionEventConfiguration.cs [hunk 24]
```diff
@@ -7,7 +7,7 @@
 using System.Resources;
 using UnityEngine;
 
-namespace Microsoft.MixedReality.Toolkit.UI.Interaction
+namespace Microsoft.MixedReality.Toolkit.Experimental.Interaction
```
Comment (julenka) [inline Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/InteractiveElement/Events/EventConfigurations/BaseInteractionEventConfiguration.cs:10]: Same here, recommend moving InteractiveElement to experimental in a separate PR to keep this a bit simpler.

File: Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/StateVisualizer/StateAnimatableProperties/AnimatableProperties/PositionOffsetStateAnimatableProperty.cs [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+        /// <summary>
+        /// The position offset added to the current position of the target object.
+        /// </summary>
+        public Vector3 PositionOffset
+        {
+            get => positionOffset;
+            set => positionOffset = value;
+        }
+
+        public PositionOffsetStateAnimatableProperty()
```
Comment (julenka) [inline Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/StateVisualizer/StateAnimatableProperties/AnimatableProperties/PositionOffsetStateAnimatableProperty.cs:27]: Double check your comments, looks like some public methods / constructors are still missing summary comments

File: Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ScaleOffsetStateAnimatableProperty.cs [hunk 26]
```diff
@@ -0,0 +1,57 @@
...
+            AnimatablePropertyName = "ScaleOffset";
+        }
+
+        public override void SetKeyFrames(AnimationClip animationClip)
+        {
+            if (Target != null)
+            {
+                string targetPath = GetTargetPath(Target);
+
+                AnimationCurve curveX = AnimationCurve.EaseInOut(0, Target.transform.localScale.x, AnimationDuration, Target.transform.localScale.x + ScaleOffset.x);
```
Comment (julenka) [inline Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/StateVisualizer/StateAnimatableProperties/AnimatableProperties/ScaleOffsetStateAnimatableProperty.cs:37]: I'm seeing a lot of repeated code in all these AnimatableProperty files. Not a huge deal, but it occured to me that you could probably refactor this to avoid repeating yourself, something like utilizing a base class, or some sort of utility method.

**Review by julenka — APPROVED** [review]
Generally seems okay now that files are moved into experimental namespace and experimental folders and I don't have a strong reason to block this going into the feature branch. I noticed that a pretty big change to also move InteractiveElement into experimental also snuck into this PR. It would be cleaner to separate out that change into a separate PR rather than cram it in here -- it will make it easier to review / look back on later.

Also, I had a suggestion to think about how to avoid lots of repeated / similar code in the animatable properties.

Finally, I noticed a few more public fields and constructors that were not commented.

File: Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/InteractiveElement/CompressableButton.cs (hunk 23 above)
Comment (CDiaz-MS) [reply Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/InteractiveElement/CompressableButton.cs]: Reverted Interactive Element changes for a separate PR

File: Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/InteractiveElement/BaseInteractiveElement.cs (hunk 22 above)
Comment (CDiaz-MS) [reply Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/InteractiveElement/BaseInteractiveElement.cs]: Reverted the changes and will modify in a separate PR.

File: Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/StateVisualizer/StateAnimatableProperties/AnimatableProperties/PositionOffsetStateAnimatableProperty.cs (hunk 25 above)
Comment (CDiaz-MS) [reply Assets/MRTK/SDK/Experimental/InteractiveElement+StateVisualizer/StateVisualizer/StateAnimatableProperties/AnimatableProperties/PositionOffsetStateAnimatableProperty.cs:30]: Added missing constructor comments
